From f135c4ba7f52ea519ed9f10efa3758fc30021164 Mon Sep 17 00:00:00 2001
From: "jiyu.yang" <jiyu.yang@amlogic.com>
Date: Wed, 12 Feb 2014 16:55:05 +0800
Subject: [PATCH 3262/5965] PD#86698 add mipi multi-channel ctrl

commit 1c7e357cb53ab46a8d0a27c76377611b0e82ed17
Author: jiyu.yang <jiyu.yang@amlogic.com>
Date:   Wed Feb 12 16:17:45 2014 +0800

    PD#86698 add mipi multi-channel ctrl

commit 67a17bc0cace51d60696c0c4e2c12ff068ea5a06
Author: jiyu.yang <jiyu.yang@amlogic.com>
Date:   Wed Feb 12 16:12:52 2014 +0800

    fix mipi-ov5647 cannt work
---
 drivers/amlogic/camera/ov5647.c           | 98 ++++++++++++++++++++++-
 include/linux/amlogic/mipi/am_mipi_csi2.h | 89 ++++++++++----------
 include/linux/amlogic/tvin/tvin_v4l2.h    |  2 +
 3 files changed, 144 insertions(+), 45 deletions(-)

diff --git a/drivers/amlogic/camera/ov5647.c b/drivers/amlogic/camera/ov5647.c
index 6d7339aa951d..ba4fefd30ac0 100755
--- a/drivers/amlogic/camera/ov5647.c
+++ b/drivers/amlogic/camera/ov5647.c
@@ -1194,8 +1194,96 @@ struct aml_camera_i2c_fig_s OV5647_preview_720P_script[] = {
 };
 
 struct aml_camera_i2c_fig_s OV5647_960P_script_mipi[] = {
-        {0x4800, 0x24},
-	{0xffff, 0xff},
+    { 0x4800, 0x24},
+    { 0x0100, 0x00},
+    { 0x0103, 0x01},
+    { 0x3035, 0x11},
+    { 0x3036, 0x46},
+    { 0x303c, 0x11},
+    { 0x3821, 0x07},
+    { 0x3820, 0x41},
+    { 0x370c, 0x03},
+    { 0x3612, 0x59},
+    { 0x3618, 0x00},
+    { 0x5000, 0x06},
+    { 0x5003, 0x08},
+    { 0x5a00, 0x08},
+    { 0x3000, 0xff},
+    { 0x3001, 0xff},
+    { 0x3002, 0xff},
+    { 0x301d, 0xf0},
+    { 0x3a18, 0x00},
+    { 0x3a19, 0xf8},
+    { 0x3c01, 0x80},
+    { 0x3b07, 0x0c},
+    { 0x380c, 0x07},
+    { 0x380d, 0x68},
+    { 0x380e, 0x03},
+    { 0x380f, 0xd8},
+    { 0x3814, 0x31},
+    { 0x3815, 0x31},
+    { 0x3708, 0x64},
+    { 0x3709, 0x52},
+    { 0x3808, 0x05},
+    { 0x3809, 0x00},
+    { 0x380a, 0x03},
+    { 0x380b, 0xc0},
+    { 0x3800, 0x00},
+    { 0x3801, 0x18},
+    { 0x3802, 0x00},
+    { 0x3803, 0x0e},
+    { 0x3804, 0x0a},
+    { 0x3805, 0x27},
+    { 0x3806, 0x07},
+    { 0x3807, 0x95},
+    { 0x3630, 0x2e},
+    { 0x3632, 0xe2},
+    { 0x3633, 0x23},
+    { 0x3634, 0x44},
+    { 0x3620, 0x64},
+    { 0x3621, 0xe0},
+    { 0x3600, 0x37},
+    { 0x3704, 0xa0},
+    { 0x3703, 0x5a},
+    { 0x3715, 0x78},
+    { 0x3717, 0x01},
+    { 0x3731, 0x02},
+    { 0x370b, 0x60},
+    { 0x3705, 0x1a},
+    { 0x3f05, 0x02},
+    { 0x3f06, 0x10},
+    { 0x3f01, 0x0a},
+    { 0x3a08, 0x01},
+    { 0x3a09, 0x27},
+    { 0x3a0a, 0x00},
+    { 0x3a0b, 0xf6},
+    { 0x3a0d, 0x04},
+    { 0x3a0e, 0x03},
+    { 0x3a0f, 0x58},
+    { 0x3a10, 0x50},
+    { 0x3a1b, 0x58},
+    { 0x3a1e, 0x50},
+    { 0x3a11, 0x60},
+    { 0x3a1f, 0x28},
+    { 0x4001, 0x02},
+    { 0x4004, 0x02},
+    { 0x4000, 0x09},
+    { 0x0100, 0x01},
+    { 0x3000, 0x00},
+    { 0x3001, 0x00},
+    { 0x3002, 0x00},
+    { 0x3017, 0xe0},
+    { 0x301c, 0xfc},
+    { 0x3636, 0x06},
+    { 0x3016, 0x08},
+    { 0x3827, 0xec},
+    { 0x4800, 0x24},
+    { 0x3018, 0x44},
+    { 0x3035, 0x21},
+    { 0x3106, 0xf5},
+    { 0x3034, 0x1a},
+    { 0x301c, 0xf8},
+	  { 0xffff, 0xff},
 };
 struct aml_camera_i2c_fig_s OV5647_preview_960P_script[] = {
 	{0x0100,0x00},  
@@ -4239,9 +4327,11 @@ static int ov5647_open(struct file *file)
     fh->height   = 480;
     fh->stream_on = 0 ;
     fh->f_flags  = file->f_flags;
+#if 0
     if( CAM_MIPI == dev->cam_info.interface){ //deprecated; this added for there is no 960p output for mipi
-        i_index = 2;
+        i_index = 3;
     }
+#endif
     /* Resets frame counters */
     dev->jiffies = jiffies;
 
@@ -4530,7 +4620,7 @@ static ssize_t cam_info_store(struct device *dev,struct device_attribute *attr,c
                 printk("substitude with %s interface\n", t->cam_info.interface?"mipi":"dvp");
         }else if ( 0 == strcmp(parm[0],"clk")){
                 t->cam_info.clk_channel = simple_strtol(parm[1],NULL,16);
-                printk("clk channel =%s\n", t->cam_info.interface?"clkB":"clkA");
+                printk("clk channel =%s\n", t->cam_info.clk_channel?"clkB":"clkA");
         }
 
         kfree(buf_orig);
diff --git a/include/linux/amlogic/mipi/am_mipi_csi2.h b/include/linux/amlogic/mipi/am_mipi_csi2.h
index 02ff69ed8d05..169654067722 100755
--- a/include/linux/amlogic/mipi/am_mipi_csi2.h
+++ b/include/linux/amlogic/mipi/am_mipi_csi2.h
@@ -9,8 +9,8 @@
  *
  *******************************************************************/
 
-#ifndef AM_MIPI_CSI2
-#define AM_MIPI_CSI2
+#ifndef __AM_MIPI_CSI2_H__
+#define __AM_MIPI_CSI2_H__
 
 #include <mach/io.h>
 #include <linux/amlogic/tvin/tvin_v4l2.h>
@@ -157,6 +157,13 @@ struct am_csi2_s{
     struct am_csi2_ops_s *ops;
 };
 
+extern int start_mipi_csi2_service(struct am_csi2_camera_para *para);
+extern int stop_mipi_csi2_service(struct am_csi2_camera_para *para);
+extern void am_mipi_csi2_init(csi_parm_t* info);
+extern void am_mipi_csi2_uninit(void);
+extern void init_am_mipi_csi2_clock(void);
+extern void cal_csi_para(csi_parm_t* info);
+
 #define MIPI_DEBUG
 #ifdef MIPI_DEBUG
 #define mipi_dbg(fmt, args...) printk(fmt,## args)
@@ -209,45 +216,45 @@ struct am_csi2_s{
 //struct device;
 //struct v4l2_device;
 
-//#define CSI2_CLK_RESET                             0x2a00
+//#define CSI2_CLK_RESET                0x2a00
 #define CSI2_CFG_CLK_ENABLE_DWC         3 
-#define CSI2_CFG_CLK_AUTO_GATE_OFF     2
-#define CSI2_CFG_CLK_ENABLE                    1
-#define CSI2_CFG_SW_RESET                       0
-
-//#define CSI2_GEN_CTRL0                             0x2a01
-#define CSI2_CFG_CLR_WRRSP                      27
-#define CSI2_CFG_DDR_EN                            26
-#define CSI2_CFG_A_BRST_NUM                   20  //25:20
-#define CSI2_CFG_A_ID                                14  //19:14
-#define CSI2_CFG_URGENT_EN                     13
-#define CSI2_CFG_DDR_ADDR_LPBK             12
-#define CSI2_CFG_BUFFER_PIC_SIZE           11
-#define CSI2_CFG_422TO444_MODE             10
-#define CSI2_CFG_INV_FIELD                        9
-#define CSI2_CFG_INTERLACE_EN                 8
-#define CSI2_CFG_FORCE_LINE_COUNT         7
-#define CSI2_CFG_FORCE_PIX_COUNT           6
-#define CSI2_CFG_COLOR_EXPAND                 5
-#define CSI2_CFG_ALL_TO_MEM                     4
+#define CSI2_CFG_CLK_AUTO_GATE_OFF      2
+#define CSI2_CFG_CLK_ENABLE             1
+#define CSI2_CFG_SW_RESET               0
+
+//#define CSI2_GEN_CTRL0                0x2a01
+#define CSI2_CFG_CLR_WRRSP              27
+#define CSI2_CFG_DDR_EN                 26
+#define CSI2_CFG_A_BRST_NUM             20  //25:20
+#define CSI2_CFG_A_ID                   14  //19:14
+#define CSI2_CFG_URGENT_EN              13
+#define CSI2_CFG_DDR_ADDR_LPBK          12
+#define CSI2_CFG_BUFFER_PIC_SIZE        11
+#define CSI2_CFG_422TO444_MODE          10
+#define CSI2_CFG_INV_FIELD              9
+#define CSI2_CFG_INTERLACE_EN           8
+#define CSI2_CFG_FORCE_LINE_COUNT       7
+#define CSI2_CFG_FORCE_PIX_COUNT        6
+#define CSI2_CFG_COLOR_EXPAND           5
+#define CSI2_CFG_ALL_TO_MEM             4
 #define CSI2_CFG_VIRTUAL_CHANNEL_EN     0  //3:0
 
-//#define CSI2_FORCE_PIC_SIZE                        0x2a02
-#define CSI2_CFG_LINE_COUNT                     16 //31:16
-#define CSI2_CFG_PIX_COUNT                       0 //15:0
-
-//#define CSI2_INTERRUPT_CTRL_STAT                   0x2a05
-#define CSI2_CFG_VS_RISE_INTERRUPT_CLR         18 
-#define CSI2_CFG_VS_FAIL_INTERRUPT_CLR         17 
-#define CSI2_CFG_FIELD_DONE_INTERRUPT_CLR  16 
-#define CSI2_CFG_VS_RISE_INTERRUPT                 2 
-#define CSI2_CFG_VS_FAIL_INTERRUPT                 1 
-#define CSI2_CFG_FIELD_DONE_INTERRUPT          0 
-
-extern int start_mipi_csi2_service(struct am_csi2_camera_para *para);
-extern int stop_mipi_csi2_service(struct am_csi2_camera_para *para);
-extern void am_mipi_csi2_init(csi_parm_t* info);
-extern void am_mipi_csi2_uninit(void);
-extern void init_am_mipi_csi2_clock(void);
-
-#endif
+//#define CSI2_FORCE_PIC_SIZE           0x2a02
+#define CSI2_CFG_LINE_COUNT             16 //31:16
+#define CSI2_CFG_PIX_COUNT              0 //15:0
+
+//#define CSI2_INTERRUPT_CTRL_STAT              0x2a05
+#define CSI2_CFG_VS_RISE_INTERRUPT_CLR          18
+#define CSI2_CFG_VS_FAIL_INTERRUPT_CLR          17
+#define CSI2_CFG_FIELD_DONE_INTERRUPT_CLR       16
+#define CSI2_CFG_VS_RISE_INTERRUPT              2
+#define CSI2_CFG_VS_FAIL_INTERRUPT              1
+#define CSI2_CFG_FIELD_DONE_INTERRUPT           0
+
+//#define MIPI_PHY_CTRL                         0x00
+#define MIPI_PHY_CFG_CHPU_TO_ANALOG             5
+#define MIPI_PHY_CFG_SHTDWN_CLK_LANE            4
+#define MIPI_PHY_CFG_SHTDWN_DATA_LANE           0
+
+#define MIPI_PHY_CFG_CLK_CHNLB_SHIFT            2
+#endif //__AM_MIPI_CSI2_H__
diff --git a/include/linux/amlogic/tvin/tvin_v4l2.h b/include/linux/amlogic/tvin/tvin_v4l2.h
index edbb142f1ff1..3fd9c4ac718a 100755
--- a/include/linux/amlogic/tvin/tvin_v4l2.h
+++ b/include/linux/amlogic/tvin/tvin_v4l2.h
@@ -746,6 +746,8 @@ typedef struct csi_parm_s {
 	unsigned hs_freq; //hz
 	unsigned urgent;
 
+        unsigned int lane_mask;
+
 	clk_channel_t clk_channel;
 	unsigned int skip_frames;
 	tvin_color_fmt_t csi_ofmt;
-- 
2.19.0

